### PR TITLE
feat: add mux tracer for native call tracer and native prestate tracers

### PR DESCRIPTION
**Pull request by 0xmountaintop**
## 1. Purpose or design rationale of this PR

1. continued work based on https://github.com/scroll-tech/go-ethereum/pull/557. background see https://github.com/scroll-tech/go-ethereum/issues/594
2. have to upgrade go version because `eth/tracers/native/prestate.go` uses some new go feature
3. then need to upgrade golangci-lint because the previous version isn't compatible with go1.20
4. use interface and move `core/trace.go` to avoid import cycle
5. suggest ignoring lint CI now, because most of the hints make little sense, and the formatting introduces a lot of diff

## 2. PR title

Your PR title must follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) (as we are doing squash merge for each PR), so it must start with one of the following [types](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type):

- [ ] build: Changes that affect the build system or external dependencies (example scopes: yarn, eslint, typescript)
- [ ] ci: Changes to our CI configuration files and scripts (example scopes: vercel, github, cypress)
- [ ] docs: Documentation-only changes
- [X] feat: A new feature
- [ ] fix: A bug fix
- [ ] perf: A code change that improves performance
- [ ] refactor: A code change that doesn't fix a bug, or add a feature, or improves performance
- [ ] style: Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
- [ ] test: Adding missing tests or correcting existing tests

## 3. Deployment tag versioning

Has the version in `params/version.go` been updated?

- [ ] This PR doesn't involve a new deployment, git tag, docker image tag, and it doesn't affect traces
- [X] Yes

## 4. Breaking change label

Does this PR have the `breaking-change` label?

- [ ] This PR is not a breaking change
- [X] Yes

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1206379435126548

### CODE REVIEW

File: rollup/tracing/tracing.go [hunk 1]
```diff
@@ -270,29 +288,44 @@ func (env *TraceEnv) getTxResult(state *state.StateDB, index int, block *types.B
...
 	l1DataFee, err := fees.CalculateL1DataFee(tx, state)
 	if err != nil {
-		return err
+		return fmt.Errorf("tracing failed: %w", err)
 	}
-	result, err := ApplyMessage(vmenv, msg, new(GasPool).AddGas(msg.Gas()), l1DataFee)
+	result, err := core.ApplyMessage(vmenv, msg, new(core.GasPool).AddGas(msg.Gas()), l1DataFee)
 	if err != nil {
-		return err
+		return fmt.Errorf("tracing failed: %w", err)
```
Comment (0xmountaintop) [inline rollup/tracing/tracing.go]: TODO: fix behaviors

Comment (0xmountaintop) [reply rollup/tracing/tracing.go]: fixed in https://github.com/scroll-tech/go-ethereum/pull/615/commits/ebd8a956f35ee9c82936844268a814be13ea553b

**Review by lightsing — DISMISSED** [review]

File: build/ci.go [hunk 2]
```diff
@@ -24,19 +24,18 @@ Usage: go run build/ci.go <command> <command flags/arguments>
...
+	test       [ -coverage ] [ packages... ]                                                    -- runs the tests
+	lint                                                                                        -- runs certain pre-selected linters
+	archive    [ -arch architecture ] [ -type zip|tar ] [ -signer key-envvar ] [ -signify key-envvar ] [ -upload dest ] -- archives build artifacts
+	importkeys                                                                                  -- imports signing keys from env
+	debsrc     [ -signer key-id ] [ -upload dest ]                                              -- creates a debian source package
+	nsis                                                                                        -- creates a Windows NSIS installer
+	aar        [ -local ] [ -sign key-id ] [-deploy repo] [ -upload dest ]                      -- creates an Android archive
+	xcode      [ -local ] [ -sign key-id ] [-deploy repo] [ -upload dest ]                      -- creates an iOS XCode framework
+	purge      [ -store blobstore ] [ -days threshold ]                                         -- purges old archives from the blobstore
```
Comment (georgehao) [inline build/ci.go:36]: seems don't need change this.

Comment (0xmountaintop) [reply build/ci.go:36]: it's changed by auto-formatting by gofmt,

File: core/types.go [hunk 3]
```diff
@@ -32,6 +32,9 @@ type Validator interface {
 	// ValidateState validates the given statedb and optionally the receipts and
 	// gas used.
 	ValidateState(block *types.Block, state *state.StateDB, receipts types.Receipts, usedGas uint64) error
+
+	// TODO: add comments
```
Comment (georgehao) [inline core/types.go]: can add comments

File: rollup/tracing/mux_tracer.go [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+import (
+	"math/big"
+	"time"
+
+	"github.com/scroll-tech/go-ethereum/common"
+	"github.com/scroll-tech/go-ethereum/core/vm"
+	_ "github.com/scroll-tech/go-ethereum/eth/tracers/native"
+)
+
+type MuxTracer struct {
```
Comment (georgehao) [inline rollup/tracing/mux_tracer.go:13]: Add comment

File: rollup/tracing/tracing.go [hunk 5]
```diff
@@ -23,6 +26,21 @@ import (
 	"github.com/scroll-tech/go-ethereum/trie/zkproof"
 )
 
+type TracerWrapper struct{}
```
Comment (georgehao) [inline rollup/tracing/tracing.go:30]: ditto, comment

**Review by georgehao — DISMISSED** [review]

File: core/types.go (hunk 3 above)
Comment (0xmountaintop) [reply core/types.go]: resolved in https://github.com/scroll-tech/go-ethereum/pull/615/commits/acf23b72b5312edadf3ff4a9a1f6a3248ab47d75

File: rollup/tracing/mux_tracer.go (hunk 4 above)
Comment (0xmountaintop) [reply rollup/tracing/mux_tracer.go:13]: resolved in https://github.com/scroll-tech/go-ethereum/pull/615/commits/acf23b72b5312edadf3ff4a9a1f6a3248ab47d75

File: rollup/tracing/tracing.go (hunk 5 above)
Comment (0xmountaintop) [reply rollup/tracing/tracing.go:30]: resolved in https://github.com/scroll-tech/go-ethereum/pull/615/commits/acf23b72b5312edadf3ff4a9a1f6a3248ab47d75

**Review by georgehao — APPROVED** [review]

**Comment by lispc** [comment]
i think after this PR, traces only add more field but not remove existed fields.  So in thoery it should be compatible with existed prover?   @HAOYUatHZ

**Comment by 0xmountaintop** [comment]
> i think after this PR, traces only add more field but not remove existed fields. So in thoery it should be compatible with existed prover? @HAOYUatHZ

yes this PR is compatible (see https://scrollco.slack.com/archives/C05UJL77FAR/p1705908908525729?thread_ts=1704082609.850259&cid=C05UJL77FAR)

but after we "disable stack", it will be incompatible and circuits need to be upgraded

**Pull request merged**